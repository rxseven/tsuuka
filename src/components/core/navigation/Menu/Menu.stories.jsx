import React from 'react';
import { storiesOf } from '@storybook/react';

import Menu from './index';

const title = 'Components/Core/Navigation/Menu';

storiesOf(title, module)
  .add('close (default)', () => <Menu />, {
    notes: 'A component is invisible by default (isOpen: false).'
  })
  .add('open', () => <Menu isOpen />);
