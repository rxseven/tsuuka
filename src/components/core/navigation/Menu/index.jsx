import PropTypes from 'prop-types';
import React, { useEffect } from 'react';
import { slide as Offcanvas } from 'react-burger-menu';
import { withRouter } from 'react-router-dom';
import styled, { css, createGlobalStyle } from 'styled-components';

import Brand from 'components/common/base/Brand';
import Hyperlink from 'components/common/base/Hyperlink';
import Icon from 'components/common/base/Icon';
import Nav from 'components/common/base/Nav';

import HTML from 'constants/elements/html';
import PATHS from 'constants/routes/paths';

const linkStyles = css`
  color: ${({ theme }) => theme.color.base};
  display: block;
  padding: 1rem 0 1rem 1rem;
  transition: background-color 0.15s ease-in-out;

  :hover {
    color: ${({ theme }) => theme.color.primary};
    background-color: rgba(0, 0, 0, 0.1);
    text-decoration: none;
  }
`;

const Frame = styled.div`
  box-shadow: none;
  outline: none;
`;

const Header = styled.div`
  align-items: center;
  border-bottom: #ccc solid 1px;
  display: flex;
  height: 46px;
  padding: 0 1rem;
`;

const Link = styled(Hyperlink)`
  ${linkStyles}
`;

const Meta = styled.div`
  bottom: 1rem;
  display: block;
  left: 1rem;
  position: absolute;
`;

const Navigation = styled.ul`
  list-style: none;
  margin: 0;
  padding: 0;
`;

const NavItem = styled.li`
  display: block;
`;

const NavLink = styled(Nav.Link)`
  ${linkStyles}
`;

const Logo = styled(Brand)`
  color: #7952b3;

  :hover {
    color: ${({ theme }) => theme.color.secondary};
  }
`;

const Overflow = createGlobalStyle`
  body.menu-open {
    overflow: hidden;
  }
`;

const Text = styled.span`
  margin-left: 1.25rem;
`;

const Version = styled(Hyperlink)`
  color: ${({ theme }) => theme.color.secondary};
  font-size: 0.75rem;
`;

const propTypes = {
  isOpen: PropTypes.bool.isRequired,
  onClose: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired
};

const styles = {
  bmMenu: {
    background: '#fff'
  },
  bmItemList: {
    padding: 0,
    position: 'relative'
  },
  bmMenuWrap: {
    maxWidth: '320px',
    top: 0,
    zIndex: 8001
  },
  bmOverlay: {
    background: 'rgba(0, 0, 0, 0.8)',
    top: 0
  }
};

const options = {
  bodyClassName: 'menu-open',
  customBurgerIcon: false,
  customCrossIcon: false,
  id: 'sidebar',
  left: true,
  outerContainerId: HTML.wrapper,
  pageWrapId: HTML.body,
  styles,
  width: '85%'
};

function Menu({ isOpen, location, onChange, onClose }) {
  // Close menu if the URL changes
  useEffect(() => {
    onClose();
  }, [location.pathname]);

  return (
    <React.Fragment>
      <Offcanvas {...options} isOpen={isOpen} onStateChange={onChange}>
        <Frame>
          <Header>
            <Logo />
          </Header>
          <Navigation>
            <NavItem>
              <NavLink exact to={PATHS.root}>
                <Icon icon="home" />
                <Text>Currency Converter</Text>
              </NavLink>
            </NavItem>
            <NavItem>
              <NavLink exact to={PATHS.about}>
                <Icon icon="info-circle" />
                <Text>About TsÅ«ka</Text>
              </NavLink>
            </NavItem>
            <NavItem>
              <NavLink exact to={PATHS.terms}>
                <Icon icon="file-alt" />
                <Text>Terms & Conditions</Text>
              </NavLink>
            </NavItem>
            <NavItem>
              <Link href="https://tsuuka.netlify.com">
                <Icon icon="star-of-life" />
                <Text>Storybook</Text>
              </Link>
            </NavItem>
            <NavItem>
              <Link href="https://github.com/rxseven/tsuuka">
                <Icon icon={['fab', 'github-alt']} />
                <Text>View on GitHub</Text>
              </Link>
            </NavItem>
          </Navigation>
          <Meta>
            <Version href="https://github.com/rxseven/tsuuka/releases/tag/v0.3.0">
              Version 0.3.0
            </Version>
          </Meta>
        </Frame>
      </Offcanvas>
      <Overflow />
    </React.Fragment>
  );
}

Menu.propTypes = propTypes;

export default withRouter(Menu);
