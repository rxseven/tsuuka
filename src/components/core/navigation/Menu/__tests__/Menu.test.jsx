import { shallow } from 'enzyme';
import React from 'react';
import renderer from 'react-test-renderer';

import Environment from 'tests/environment';
import Menu from '../index';

describe('<Menu />', () => {
  // Arrange
  const component = (
    <Environment>
      <Menu />
    </Environment>
  );

  describe('Unit tests', () => {
    it('should render without crashing', () => {
      // Act
      const wrapper = shallow(component);

      // Assert
      expect(wrapper).toBeDefined();
    });
  });

  describe('Snapshot tests', () => {
    it('should render correctly', () => {
      // Act
      const tree = renderer.create(component).toJSON();

      // Assert
      expect(tree).toMatchSnapshot();
    });
  });
});
