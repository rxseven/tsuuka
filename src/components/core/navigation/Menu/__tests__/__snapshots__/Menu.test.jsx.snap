// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Menu /> Snapshot tests should render correctly 1`] = `
<div>
  <div
    className="bm-overlay"
    onClick={[Function]}
    style={
      Object {
        "MozTransform": "translate3d(100%, 0, 0)",
        "MsTransform": "translate3d(100%, 0, 0)",
        "OTransform": "translate3d(100%, 0, 0)",
        "WebkitTransform": "translate3d(100%, 0, 0)",
        "background": "rgba(0, 0, 0, 0.8)",
        "height": "100%",
        "opacity": 0,
        "position": "fixed",
        "top": 0,
        "transform": "translate3d(100%, 0, 0)",
        "transition": "opacity 0.3s, transform 0s 0.3s",
        "width": "100%",
        "zIndex": 1000,
      }
    }
  />
  <div
    className="bm-menu-wrap"
    id="sidebar"
    style={
      Object {
        "MozTransform": "translate3d(-100%, 0, 0)",
        "MsTransform": "translate3d(-100%, 0, 0)",
        "OTransform": "translate3d(-100%, 0, 0)",
        "WebkitTransform": "translate3d(-100%, 0, 0)",
        "height": "100%",
        "maxWidth": "320px",
        "position": "fixed",
        "right": "inherit",
        "top": 0,
        "transform": "translate3d(-100%, 0, 0)",
        "transition": "all 0.5s",
        "width": "85%",
        "zIndex": 8001,
      }
    }
  >
    <div
      className="bm-menu"
      style={
        Object {
          "background": "#fff",
          "boxSizing": "border-box",
          "height": "100%",
          "overflow": "auto",
        }
      }
    >
      <nav
        className="bm-item-list"
        style={
          Object {
            "height": "100%",
            "padding": 0,
            "position": "relative",
          }
        }
      >
        <div
          className="bm-item sc-htpNat bcNAOR"
          style={
            Object {
              "display": "block",
            }
          }
          tabIndex={-1}
        >
          <div
            className="sc-bxivhb joJJRO"
          >
            <a
              className="sc-dnqmqq cxTrtM sc-bdVaJa gRJttW"
              href="/"
              onClick={[Function]}
              title="Currency Converter"
            >
              通貨換算ツール
            </a>
          </div>
          <ul
            className="sc-bZQynM hRyDTQ"
          >
            <li
              className="sc-gzVnrw kXANwb"
            >
              <a
                aria-current="page"
                className="sc-htoDjs ebvccb sc-bwzfXH dRlolo active"
                data-light={false}
                href="/"
                onClick={[Function]}
                style={Object {}}
              >
                <svg
                  aria-hidden="true"
                  className="svg-inline--fa fa-home fa-w-18 "
                  data-icon="home"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  style={Object {}}
                  viewBox="0 0 576 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M280.37 148.26L96 300.11V464a16 16 0 0 0 16 16l112.06-.29a16 16 0 0 0 15.92-16V368a16 16 0 0 1 16-16h64a16 16 0 0 1 16 16v95.64a16 16 0 0 0 16 16.05L464 480a16 16 0 0 0 16-16V300L295.67 148.26a12.19 12.19 0 0 0-15.3 0zM571.6 251.47L488 182.56V44.05a12 12 0 0 0-12-12h-56a12 12 0 0 0-12 12v72.61L318.47 43a48 48 0 0 0-61 0L4.34 251.47a12 12 0 0 0-1.6 16.9l25.5 31A12 12 0 0 0 45.15 301l235.22-193.74a12.19 12.19 0 0 1 15.3 0L530.9 301a12 12 0 0 0 16.9-1.6l25.5-31a12 12 0 0 0-1.7-16.93z"
                    fill="currentColor"
                    style={Object {}}
                  />
                </svg>
                <span
                  className="sc-iwsKbI gVMPMw"
                >
                  Currency Converter
                </span>
              </a>
            </li>
            <li
              className="sc-gzVnrw kXANwb"
            >
              <a
                aria-current={null}
                className="sc-htoDjs ebvccb sc-bwzfXH dRlolo"
                data-light={false}
                href="/about"
                onClick={[Function]}
              >
                <svg
                  aria-hidden="true"
                  className="svg-inline--fa fa-info-circle fa-w-16 "
                  data-icon="info-circle"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  style={Object {}}
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 110c23.196 0 42 18.804 42 42s-18.804 42-42 42-42-18.804-42-42 18.804-42 42-42zm56 254c0 6.627-5.373 12-12 12h-88c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h12v-64h-12c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h64c6.627 0 12 5.373 12 12v100h12c6.627 0 12 5.373 12 12v24z"
                    fill="currentColor"
                    style={Object {}}
                  />
                </svg>
                <span
                  className="sc-iwsKbI gVMPMw"
                >
                  About Tsūka
                </span>
              </a>
            </li>
            <li
              className="sc-gzVnrw kXANwb"
            >
              <a
                aria-current={null}
                className="sc-htoDjs ebvccb sc-bwzfXH dRlolo"
                data-light={false}
                href="/terms"
                onClick={[Function]}
              >
                <svg
                  aria-hidden="true"
                  className="svg-inline--fa fa-file-alt fa-w-12 "
                  data-icon="file-alt"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  style={Object {}}
                  viewBox="0 0 384 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M224 136V0H24C10.7 0 0 10.7 0 24v464c0 13.3 10.7 24 24 24h336c13.3 0 24-10.7 24-24V160H248c-13.2 0-24-10.8-24-24zm64 236c0 6.6-5.4 12-12 12H108c-6.6 0-12-5.4-12-12v-8c0-6.6 5.4-12 12-12h168c6.6 0 12 5.4 12 12v8zm0-64c0 6.6-5.4 12-12 12H108c-6.6 0-12-5.4-12-12v-8c0-6.6 5.4-12 12-12h168c6.6 0 12 5.4 12 12v8zm0-72v8c0 6.6-5.4 12-12 12H108c-6.6 0-12-5.4-12-12v-8c0-6.6 5.4-12 12-12h168c6.6 0 12 5.4 12 12zm96-114.1v6.1H256V0h6.1c6.4 0 12.5 2.5 17 7l97.9 98c4.5 4.5 7 10.6 7 16.9z"
                    fill="currentColor"
                    style={Object {}}
                  />
                </svg>
                <span
                  className="sc-iwsKbI gVMPMw"
                >
                  Terms & Conditions
                </span>
              </a>
            </li>
            <li
              className="sc-gzVnrw kXANwb"
            >
              <a
                className="sc-ifAKCX fhesyB"
                href="https://tsuuka.netlify.com"
                rel="noopener noreferrer"
                target="_blank"
              >
                <svg
                  aria-hidden="true"
                  className="svg-inline--fa fa-star-of-life fa-w-15 "
                  data-icon="star-of-life"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  style={Object {}}
                  viewBox="0 0 480 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M471.99 334.43L336.06 256l135.93-78.43c7.66-4.42 10.28-14.2 5.86-21.86l-32.02-55.43c-4.42-7.65-14.21-10.28-21.87-5.86l-135.93 78.43V16c0-8.84-7.17-16-16.01-16h-64.04c-8.84 0-16.01 7.16-16.01 16v156.86L56.04 94.43c-7.66-4.42-17.45-1.79-21.87 5.86L2.15 155.71c-4.42 7.65-1.8 17.44 5.86 21.86L143.94 256 8.01 334.43c-7.66 4.42-10.28 14.21-5.86 21.86l32.02 55.43c4.42 7.65 14.21 10.27 21.87 5.86l135.93-78.43V496c0 8.84 7.17 16 16.01 16h64.04c8.84 0 16.01-7.16 16.01-16V339.14l135.93 78.43c7.66 4.42 17.45 1.8 21.87-5.86l32.02-55.43c4.42-7.65 1.8-17.43-5.86-21.85z"
                    fill="currentColor"
                    style={Object {}}
                  />
                </svg>
                <span
                  className="sc-iwsKbI gVMPMw"
                >
                  Storybook
                </span>
              </a>
            </li>
            <li
              className="sc-gzVnrw kXANwb"
            >
              <a
                className="sc-ifAKCX fhesyB"
                href="https://github.com/rxseven/tsuuka"
                rel="noopener noreferrer"
                target="_blank"
              >
                <svg
                  aria-hidden="true"
                  className="svg-inline--fa fa-github-alt fa-w-15 "
                  data-icon="github-alt"
                  data-prefix="fab"
                  focusable="false"
                  role="img"
                  style={Object {}}
                  viewBox="0 0 480 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M186.1 328.7c0 20.9-10.9 55.1-36.7 55.1s-36.7-34.2-36.7-55.1 10.9-55.1 36.7-55.1 36.7 34.2 36.7 55.1zM480 278.2c0 31.9-3.2 65.7-17.5 95-37.9 76.6-142.1 74.8-216.7 74.8-75.8 0-186.2 2.7-225.6-74.8-14.6-29-20.2-63.1-20.2-95 0-41.9 13.9-81.5 41.5-113.6-5.2-15.8-7.7-32.4-7.7-48.8 0-21.5 4.9-32.3 14.6-51.8 45.3 0 74.3 9 108.8 36 29-6.9 58.8-10 88.7-10 27 0 54.2 2.9 80.4 9.2 34-26.7 63-35.2 107.8-35.2 9.8 19.5 14.6 30.3 14.6 51.8 0 16.4-2.6 32.7-7.7 48.2 27.5 32.4 39 72.3 39 114.2zm-64.3 50.5c0-43.9-26.7-82.6-73.5-82.6-18.9 0-37 3.4-56 6-14.9 2.3-29.8 3.2-45.1 3.2-15.2 0-30.1-.9-45.1-3.2-18.7-2.6-37-6-56-6-46.8 0-73.5 38.7-73.5 82.6 0 87.8 80.4 101.3 150.4 101.3h48.2c70.3 0 150.6-13.4 150.6-101.3zm-82.6-55.1c-25.8 0-36.7 34.2-36.7 55.1s10.9 55.1 36.7 55.1 36.7-34.2 36.7-55.1-10.9-55.1-36.7-55.1z"
                    fill="currentColor"
                    style={Object {}}
                  />
                </svg>
                <span
                  className="sc-iwsKbI gVMPMw"
                >
                  View on GitHub
                </span>
              </a>
            </li>
          </ul>
          <div
            className="sc-EHOje kTJQym"
          >
            <a
              className="sc-gZMcBi cYeVZO"
              href="https://github.com/rxseven/tsuuka/releases/tag/v0.1.0"
              rel="noopener noreferrer"
              target="_blank"
            >
              Version 0.1.0
            </a>
          </div>
        </div>
      </nav>
    </div>
  </div>
</div>
`;
